module/zfs/spa_errlog.c

```diff
@@ -138,81 +570,279 @@ spa_get_errlog_size(spa_t *spa)
-
-	mutex_enter(&spa->spa_errlist_lock);
-	total += avl_numnodes(&spa->spa_errlist_last);
-	total += avl_numnodes(&spa->spa_errlist_scrub);
-	mutex_exit(&spa->spa_errlist_lock);
-
+	if (!spa_feature_is_enabled(spa, SPA_FEATURE_HEAD_ERRLOG)) {
+		mutex_enter(&spa->spa_errlog_lock);
+		if (spa->spa_errlog_scrub != 0 &&
+		    zap_count(spa->spa_meta_objset, spa->spa_errlog_scrub,
+		    &count) == 0)
+			total += count;
+
+		if (spa->spa_errlog_last != 0 && !spa->spa_scrub_finished &&
+		    zap_count(spa->spa_meta_objset, spa->spa_errlog_last,
+		    &count) == 0)
+			total += count;
+		mutex_exit(&spa->spa_errlog_lock);
+
+		mutex_enter(&spa->spa_errlist_lock);
+		total += avl_numnodes(&spa->spa_errlist_last);
+		total += avl_numnodes(&spa->spa_errlist_scrub);
+		mutex_exit(&spa->spa_errlist_lock);
+	} else {
+#ifdef _KERNEL
+		mutex_enter(&spa->spa_errlog_lock);
+		total += get_errlog_size(spa, spa->spa_errlog_last);
+		total += get_errlog_size(spa, spa->spa_errlog_scrub);
+		mutex_exit(&spa->spa_errlog_lock);
+
+		mutex_enter(&spa->spa_errlist_lock);
+		total += get_errlist_size(spa, &spa->spa_errlist_last);
+		total += get_errlist_size(spa, &spa->spa_errlist_scrub);
+		mutex_exit(&spa->spa_errlist_lock);
+#endif
+	}
 	return (total);
 }
 
-#ifdef _KERNEL
-static int
-process_error_log(spa_t *spa, uint64_t obj, void *addr, size_t *count)
+/*
+ * This function sweeps through an on-disk error log and stores all bookmarks
+ * as error bookmarks in a new ZAP object. At the end we discard the old one,
+ * and spa_update_errlog() will set the spa's on-disk error log to new ZAP
+ * object.
+ */
+static void
+sync_upgrade_errlog(spa_t *spa, uint64_t spa_err_obj, uint64_t *newobj,
+    dmu_tx_t *tx)
 {
 	zap_cursor_t zc;
 	zap_attribute_t za;
 	zbookmark_phys_t zb;
+	uint64_t count;
 
-	if (obj == 0)
-		return (0);
+	/*
+	 * If we cannnot perform the upgrade we should clear the old on-disk
+	 * error logs.
+	 */
+	if (zap_count(spa->spa_meta_objset, spa_err_obj, &count) != 0) {
+		VERIFY0(dmu_object_free(spa->spa_meta_objset, spa_err_obj, tx));
+		*newobj = zap_create(spa->spa_meta_objset, DMU_OT_ERROR_LOG,
+		    DMU_OT_NONE, 0, tx);
+		return;
+	}
 
-	for (zap_cursor_init(&zc, spa->spa_meta_objset, obj);
+	*newobj = zap_create(spa->spa_meta_objset, DMU_OT_ERROR_LOG,
+	    DMU_OT_NONE, 0, tx);
+
+	for (zap_cursor_init(&zc, spa->spa_meta_objset, spa_err_obj);
 	    zap_cursor_retrieve(&zc, &za) == 0;
 	    zap_cursor_advance(&zc)) {
+		if (spa_upgrade_errlog_limit != 0 &&
+		    zc.zc_cd == spa_upgrade_errlog_limit)
+			break;
+
+		name_to_bookmark(za.za_name, &zb);
 
-		if (*count == 0) {
-			zap_cursor_fini(&zc);
-			return (SET_ERROR(ENOMEM));
+		zbookmark_err_phys_t zep;
+		zep.zb_object = zb.zb_object;
+		zep.zb_level = zb.zb_level;
+		zep.zb_blkid = zb.zb_blkid;
+
+		/*
+		 * We cannot use get_head_and_birth_txg() because it will
+		 * acquire the pool config lock, which we already have. In case
+		 * of an error we simply continue.
```

Comment:
Yes, if `sync_upgrade_errlog()` cannot figure out the dataset, the objset, and the dnode of the object then it will not upgrade the old error entry and will discard it at the end.

This may happen in two cases:
1) The object does not actually belong to any dataset anymore (eg. a file was deleted and is not present in any snapshot or clone).
2) The object belongs to a dataset but `sync_upgrade_errlog()` cannot figure that out (because `dsl_dataset_hold_obj()`, `dmu_objset_from_ds()` or `dnode_hold()` error out).

For the zap entry we actually need to know only the dataset the block belongs to. The objset and the dnode are used to determine the block pointer and then if `BP_IS_HOLE()`. If we fail to retrieve the objset and the dnode (and subsequently fail to determine `BP_IS_HOLE()`) would it be reasonable to err on the side of caution and still create a zap entry (presuming that we know the dataset)?

Do you have another suggestion?

Reply:
Hm, good question. I lean toward the current semantic of dropping the error if we can't "verify" it here, otherwise we might get spurious errors that we can't ever resolve? An error here could indicate further corruption/errors and if this error is truly persistent, it should re-appear after a scan.

Reply:
Seems reasonable to me, I will leave it as it is and expand the comment.